### ISSUE TITLE
`redundant_closure` doesn't handle `Fn` 

### ISSUE THREAD

**Issue by mladedav** [rust-lang/rust-clippy#12853]
### Summary

When a closure `|val| fun(val)` is used where `fun` is `Fn`, the suggestion is to use it directly as `fun`. However, that may not compile and the correct suggestion is `&fun`.

### Lint Name

redundat_closure

### Reproducer

I tried this code:

```rust
#![warn(clippy::redundant_closure)]
#![allow(dead_code)]

fn foo<F: Fn(u32) -> u32 + 'static>(f: F) -> Option<u32> {
    let x = Box::new(move || None.map(|x| f(x)));

    x()
}
```

I saw this happen:

```
    Checking playground v0.0.1 (/playground)
warning: redundant closure
 --> src/lib.rs:5:39
  |
5 |     let x = Box::new(move || None.map(|x| f(x)));
  |                                       ^^^^^^^^ help: replace the closure with the function itself: `f`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure
note: the lint level is defined here
 --> src/lib.rs:1:9
  |
1 | #![warn(clippy::redundant_closure)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^

warning: `playground` (lib) generated 1 warning (run `cargo clippy --fix --lib -p playground` to apply 1 suggestion)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.62s
```

and if I apply that fix, I get:

```   Compiling playground v0.0.1 (/playground)
error[E0525]: expected a closure that implements the `Fn` trait, but this closure only implements `FnOnce`
 --> src/lib.rs:5:22
  |
5 |     let x = Box::new(move || None.map(f));
  |                      ^^^^^^^          - closure is `FnOnce` because it moves the variable `f` out of its environment
  |                      |
  |                      this closure implements `FnOnce`, not `Fn`
6 |
7 |     x()
  |     --- the requirement to implement `Fn` derives from here
  |
  = note: required for `Box<{closure@src/lib.rs:5:22: 5:29}>` to implement `Fn<()>`

For more information about this error, try `rustc --explain E0525`.
error: could not compile `playground` (lib) due to 1 previous error
```

I expected to see this happen (note the ampersand, that's what's missing):
```
...

5 |     let x = Box::new(move || None.map(|x| f(x)));
  |                                       ^^^^^^^^ help: replace the closure with the function itself: `&f`
...
```

### Version

```text
1.78.0 (playground)
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by J-ZhengLi** [comment]
looks like another adjustment problem?

@rustbot claim
